### ISSUE TITLE
DataQuery passes classes that don’t exist to ClassInfo::subclassesFor()

### ISSUE THREAD

**Issue by kinglozzer** [silverstripe/silverstripe-framework#4474]
[See here](https://github.com/silverstripe/silverstripe-framework/blob/ffbeac6b7d3ed1a6e02a150573ee4b2f9251cf9c/model/DataQuery.php#L208). If there is a collision in a join table (e.g. `SiteTree.Sort` colliding with `Page_ManyManyRelation.Sort`), then `Page_ManyManyRelation` is passed to `ClassInfo::subclassesFor()`, even though no PHP class exists for it.

Simplest fix is to amend the line linked above to:
```php
if (class_exists($collisionBase)) {
    $collisionClasses = ClassInfo::subclassesFor($collisionBase);
} else {
    $collisionClasses = array($collisionBase => $collisionBase);
}
```

**Comment by kinglozzer** [comment]
Now I look at this again, I don’t see how this has ever worked. Perhaps there’s another regression I’m missing... will dig further tomorrow.

**Comment by kinglozzer** [comment]
Okay this is seems to be a valid bug. The SQL that _was_ produced is valid, but doesn’t really make any sense. Pseudo-code until I get time to write up a test case:

```php
class Page extends SiteTree {
    private static $many_many = array(
        'Slides' => 'Slide'
    );

    private static $many_many_extraFields = array(
        'Slides' => array(
            'Sort' => 'Int'
        )
   );

    public function getCMSFields() {
        $fields = parent::getCMSFields();
        $fields->addFieldToTab('Root.Slides', GridField::create('Slides', 'Slides', $this->Slides(), GridFieldConfig_RecordEditor::create()));
        return $fields;
    }
}

class Slide extends DataObject {
    private static $db = array(
        'Title' => 'Varchar'
    );

    private static $belongs_many_many = array(
        'Pages' => 'Page'
    );
}
```

The error happens when you then try to edit a slide. Seems to be triggered by a `GridField` component, as adding `$fields->removeByName('Pages');` to the slide stops it happening (GF is not the cause of the issue, just what’s triggering it in this case).

Tacking on showqueries shows that the _old_ SQL produced included:

```sql
SELECT DISTINCT CASE
  WHEN "SiteTree"."ClassName" IN ('Page_Slides')
    THEN "Page_Slides"."Sort"
  WHEN "SiteTree"."ClassName" IN ('SiteTree', 'Page', 'HomePage', 'ErrorPage', 'RedirectorPage', 'VirtualPage')
    THEN "SiteTree"."Sort"
  WHEN "SiteTree"."ClassName" IN ('SiteTree', 'Page', 'HomePage', 'ErrorPage', 'RedirectorPage', 'VirtualPage')
    THEN "SiteTree"."Sort"
  WHEN "SiteTree"."ClassName" IN ('SiteTree', 'Page', 'HomePage', 'ErrorPage', 'RedirectorPage', 'VirtualPage')
    THEN "SiteTree"."Sort"
  ELSE NULL
END AS "Sort",

"SiteTree"."ClassName", "SiteTree"."Created", "SiteTree"."LastEdited", "SiteTree"."URLSegment", "SiteTree"."Title", "SiteTree"."MenuTitle", "SiteTree"."Content", "SiteTree"."MetaDescription", "SiteTree"."ExtraMeta", "SiteTree"."ShowInMenus", "SiteTree"."ShowInSearch", "SiteTree"."HasBrokenFile", "SiteTree"."HasBrokenLink", "SiteTree"."ReportClass", "SiteTree"."CanViewType", "SiteTree"."CanEditType", "SiteTree"."GridLayout", "SiteTree"."MetaTitle", "SiteTree"."Version", "SiteTree"."ParentID", "SiteTree"."ID", CASE WHEN "SiteTree"."ClassName" IS NOT NULL THEN "SiteTree"."ClassName" ELSE 'SiteTree' END AS "RecordClassName" FROM "SiteTree" INNER JOIN "Page_Slides" ON "Page_Slides"."PageID" = "SiteTree"."ID" WHERE ("Page_Slides"."SlideID" = '2') AND ("SiteTree"."ClassName" IN ('Page','HomePage','ErrorPage','RedirectorPage','VirtualPage')) ORDER BY "SiteTree"."Sort" ASC LIMIT 15
```

Obviously the class name can never be `Page_Slides`, so it shouldn’t be including that. Would probably be a good idea to tidy up the duplicate `WHEN` cases too.

**Comment by kinglozzer** [comment]
Adding this above the line highlighted in the original issue above to this fixes it:

```php
if (!class_exists($collisionBase)) {
	continue;
}
```

I’m really struggling to capture this in a test case, though...

**Comment by kinglozzer** [comment]
Fixed in #4497 / #4482
